File: conformance/run-1/components/channels.py
import asyncio

from jiggy.flow import OperationError, handle


async def run(context):
    channel = await context.channel()
    work = asyncio.create_task(context.call_capability(
        operation_id="answer", slot="worker", method="run", input=context.input,
        channels={"events": channel.send},
    ))
    values = []
    complete = True
    try:
        async for value in channel.receive:
            values.append(value)
    except OperationError as error:
        if error.code != "LAGGED":
            await asyncio.gather(work, return_exceptions=True)
            raise
        complete = False
    return {"outcome": "done", "output": {"complete": complete, "values": values, "work": await work}}


handle(run)

File: conformance/run-1/components/channels.ts
import { handle, type JsonValue, OperationError } from '../../../packages/flow-sdk/src/index'

await handle(async (run) => {
  const channel = await run.channel()
  const work = run.call({
    operationId: 'answer',
    slot: 'worker',
    input: run.input,
    channels: { events: channel.send },
  })
  const values: JsonValue[] = []
  let complete = true
  try {
    for await (const value of channel.receive) values.push(value)
  } catch (error) {
    if (!(error instanceof OperationError) || error.code !== 'LAGGED') {
      await Promise.allSettled([work])
      throw error
    }
    complete = false
  }
  return { outcome: 'done', output: { complete, values, work: await work } }
})
